test/unit/org/apache/cassandra/transport/TransportTest.java

```diff
@@ -0,0 +1,108 @@
+ * limitations under the License.
+ */
+package org.apache.cassandra.transport;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+
+import org.apache.cassandra.concurrent.Stage;
+import org.apache.cassandra.config.CassandraRelevantProperties;
+import org.apache.cassandra.cql3.CQLTester;
+import org.apache.cassandra.cql3.QueryOptions;
+import org.apache.cassandra.cql3.QueryProcessor;
+import org.apache.cassandra.cql3.statements.BatchStatement;
+import org.apache.cassandra.transport.messages.BatchMessage;
+import org.apache.cassandra.transport.messages.ExecuteMessage;
+import org.apache.cassandra.transport.messages.PrepareMessage;
+import org.apache.cassandra.transport.messages.QueryMessage;
+import org.apache.cassandra.transport.messages.ResultMessage;
+
+public class TransportTest extends CQLTester
+{
+    @After
+    public void dropCreatedTable()
+    {
+        try
+        {
+            QueryProcessor.executeOnceInternal("DROP TABLE " + KEYSPACE + ".atable");
+        }
+        catch (Throwable t)
+        {
+            // ignore
+        }
+    }
+
+    @Test
+    public void testAsyncTransport() throws Throwable
+    {
+        CassandraRelevantProperties.NATIVE_TRANSPORT_ASYNC_READ_WRITE_ENABLED.setBoolean(true);
+        try
+        {
+            // Async native transport doubles the number of read/write/batch tasks
+            doTestTransport(2);
+        }
+        finally
+        {
+            CassandraRelevantProperties.NATIVE_TRANSPORT_ASYNC_READ_WRITE_ENABLED.setBoolean(false);
+        }
+    }
+
+    private void doTestTransport(long expectedTasks) throws Throwable
+    {
+        SimpleClient client = new SimpleClient(nativeAddr.getHostAddress(), nativePort);
+        try
+        {
+            requireNetwork();
+
+            client.connect(false);
+
+            QueryMessage createMessage = new QueryMessage("CREATE TABLE " + KEYSPACE + ".atable (pk int PRIMARY KEY, v text)", QueryOptions.DEFAULT);
+            PrepareMessage prepareMessage = new PrepareMessage("SELECT * FROM " + KEYSPACE + ".atable", null);
+
+            Message.Response createResponse = client.execute(createMessage);
+            ResultMessage.Prepared prepareResponse = (ResultMessage.Prepared) client.execute(prepareMessage);
+
+            ExecuteMessage executeMessage = new ExecuteMessage(prepareResponse.statementId, prepareResponse.resultMetadataId, QueryOptions.DEFAULT);
+            Message.Response executeResponse = client.execute(executeMessage);
+            Assert.assertEquals(expectedTasks, Stage.READ.executor().getCompletedTaskCount());
+
+            QueryMessage readMessage = new QueryMessage("SELECT * FROM " + KEYSPACE + ".atable", QueryOptions.DEFAULT);
+            Message.Response readResponse = client.execute(readMessage);
+            Assert.assertEquals(expectedTasks * 2, Stage.READ.executor().getCompletedTaskCount());
```

Comment:
`Assert.assertEquals(2 * numTasksMultiplier, Stage.READ.executor().getCompletedTaskCount());`

because we sent 2 queries, both handled in an async way

Please consider sending more queries or a mix of sync/async queries because the value `2` for both the multiplier and the number of queries is confusing

Reply:
> Please consider sending more queries or a mix of sync/async queries

What does that mean?

Anyway, as I said I will at least add javadocs to clarify.

Reply:
each assert in test was done after sending 2 queries. This causes the assertion condition be `2 * 2`
It's easy to mix up 2 meaning 2 queries and 2 meaning the multiplier.

Sending 3 queries would make it easier to understand what each number means. There would be `3 * multiplier`
Sending, say 3 async and 1 sync query would also yield a better assertion condition, like: `1 + 3 * multiplier`

but it's just a detail